### PR TITLE
NULL check in UDPEndPoint Sockets SendMsgImpl

### LINKED ISSUE #22552: [BUG] Potential null-pointer usage when running out of memory/buffers during minmdns reply sending
### Reproduction steps

```bash
This is an issue for #22432

Minmdns sends buffer clones over udp, to allow for multiple interfaces. The result of CloneData is never checked, potentially resulting in nullptr being used
```

### Bug prevalence

Only out of memory (not easily reproducible)

### GitHub hash of the SDK that was being used

N/A

### Platform

core

### Platform Version(s)

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by s-mcclain**
#### Problem
When CHIP Server sends UDP message to UDPEndPoint, the Server attempts `CloneData()` to clone the buffer for sending to multiple interfaces. In the rare circumstance where CloneData allocation fails, it returns an empty handle where 'this' is NULL. This causes a segmentation fault inside SendMsgImpl where it attempts to deference the empty handle without checking for null first.

#### Change overview
* Add 'VerifyOrReturnError' where if msg.IsNull, the function returns CHIP_ERROR_NO_MEMORY.

Fixes #22552

#### Testing
How was this tested? (at least one bullet point required)
* Manually tested, performing 800+ toggles, with limited memory available in system to verify no segmentation fault occurs.

### CODE REVIEW

File: src/inet/UDPEndPointImplSockets.cpp [hunk 1]
```diff
@@ -281,6 +281,9 @@ CHIP_ERROR UDPEndPointImplSockets::SendMsgImpl(const IPPacketInfo * aPktInfo, Sy
     // Ensure the destination address type is compatible with the endpoint address type.
     VerifyOrReturnError(mAddrType == aPktInfo->DestAddress.Type(), CHIP_ERROR_INVALID_ARGUMENT);
 
+    // Ensure packet buffer is not null
+    VerifyOrReturnError(!msg.IsNull(), CHIP_ERROR_NO_MEMORY);
```
Comment (tcarmelveilleux) [inline src/inet/UDPEndPointImplSockets.cpp:285]: Could you please move this to the very beginning of the checks? We should not validate `GetSocket` outcome if the message is not going to be used.

**Review by tcarmelveilleux — APPROVED** [review]

File: src/inet/UDPEndPointImplSockets.cpp (hunk 1 above)
Comment (s-mcclain) [reply src/inet/UDPEndPointImplSockets.cpp:285]: You got it!

File: src/inet/UDPEndPointImplSockets.cpp [hunk 2]
```diff
@@ -274,6 +274,9 @@ CHIP_ERROR UDPEndPointImplSockets::ListenImpl()
 
 CHIP_ERROR UDPEndPointImplSockets::SendMsgImpl(const IPPacketInfo * aPktInfo, System::PacketBufferHandle && msg)
 {
+    // Ensure packet buffer is not null
+    VerifyOrReturnError(!msg.IsNull(), CHIP_ERROR_NO_MEMORY);
```
Comment (andy31415) [inline src/inet/UDPEndPointImplSockets.cpp:278]: Should this check also be done in src/inet/UDPEndPoint.cpp ?

Here, null message means 'invalid argument' not 'no memory'. Allocating it in UDPEndpoint or any other caller would be no memory.

Comment (andy31415) [reply src/inet/UDPEndPointImplSockets.cpp:278]: Looking at call stacks, it looks like SendMsg also just passes a buffer around.

I believe 'NO_MEMORY' should be returned as a result of an allocation failure. Here it seems just the packet is null in an input argument that should never be null. that is `CHIP_ERROR_INVALID_ARGUMENT`

Comment (bzbarsky-apple) [inline src/inet/UDPEndPointImplSockets.cpp:278]: A sanity check here is probably ok, but why is someone passing in null?  Is this the `CloneData` calls in `src/lib/dnssd/minimal_mdns/Server.cpp`?  If so, _those_ should be checking that the clone succeeded...

Comment (s-mcclain) [reply src/inet/UDPEndPointImplSockets.cpp:278]: That's correct, @bzbarsky-apple . It's CloneData which is why it is possible to see the msg being null.

So you'd have a preference of having the check inside `UDPEndPoint::SendMsg()`?

Comment (bzbarsky-apple) [reply src/inet/UDPEndPointImplSockets.cpp:278]: I'd prefer that whoever does the `CloneData` checks the result before passing it around.  They are the ones who are in the best position to decide what to do if the clone fails.

Comment (s-mcclain) [reply src/inet/UDPEndPointImplSockets.cpp:278]: Sure I can agree with that argument to change the return status from being NO_MEMORY to being invalid argument.

Before I change the status code, do you agree with the conversation below with @bzbarsky-apple  about moving this check from the SendMsgImpl to being inside the UDPEndPoint::SendMsg function?

Comment (s-mcclain) [reply src/inet/UDPEndPointImplSockets.cpp:278]: That's a fair point. So we can keep this check as sanity. The caller is from the ServerBase::BroadcastImpl() inside Server.cpp. Will get another commit up with said recommendations.

Comment (s-mcclain) [reply src/inet/UDPEndPointImplSockets.cpp]: @andy31415  I've changed the return value from CHIP_ERROR_NO_MEMORY --> CHIP_ERROR_INVALID_ARGUMENT.

Comment (s-mcclain) [reply src/inet/UDPEndPointImplSockets.cpp]: I've added the checks into the MinimalMdns server in the latest commit, please feel free to review.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by andy31415 — APPROVED** [review]

**Pull request merged**